Feature/error reporting db setup

cadecameron opened the pull request:
Hey @Rovack ,

Take a look at the following pull request, which sets-up the backend error reporting table/schema/controllers/validation -- feel free to comment, give feedback, etc.

I'll probably need to discuss with you the flow when actual db operations occur (using the data-changers.js helpers, etc), and how this interacts with Lambda, etc.

Rovack commented on src/controllers/error-reports.js:
```diff
@@ -0,0 +1,105 @@
+      await Joi.validate(req, errorReportSchemas.get, { allowUnknown: true });
+
+      // Authentication of administrative user
+      if (!req.userIsAdmin) {
```
For an action like this, I don't think we have to require admin; a logged-in user should be sufficient.
Looks like you already added the middleware for that to the route, so this code can just be removed.

Rovack commented on src/models/error-report.js:
```diff
@@ -0,0 +1,36 @@
+      primaryKey: true,
+      defaultValue: DataTypes.UUIDV4,
+    },
+    services: {
```
IIRC one of the checkbox options is actually reporting about the location itself, not a specific service.
How would that be reflected in the schema?

cadecameron replied:
Hmm...
It's a checkbox for general location information, you're right.

I guess a few ways of modelling this would be to:

1. Have `services` be null (i.e. no named services were specifically selected), which would represent the error report relates to location as a whole.
1. Make `services` be an array of strings (rather than UUIDs), with one of the strings being 'location' if it was selected. Not sure how this would impact on SQL query syntax complexity, however.

I'd probably lean towards the first option to keep things simple? Open to suggestions though.

Rovack replied:
Yeah, definitely option 4.
As you point out, option 1 doesn't work with the current designs, 2 would complicate queries and prevent certain types of constraints, and 3 is the same as 4 except it introduces unnecessary JSON complexity.

cadecameron replied:
Sounds good @Rovack . I'll make the necessary changes to implement option 4.

Rovack commented on src/models/error-report.js:
```diff
@@ -0,0 +1,36 @@
+    content: DataTypes.TEXT,
+    posted_by: DataTypes.TEXT, // Currently not collected from React frontend
+    contact_info: DataTypes.TEXT, // Currently not collected from React frontend
+    hidden: DataTypes.BOOLEAN, // Not currently required (table not public-facing)
```
As discussed, `hidden` isn't really relevant for error reports.

Rovack commented on src/controllers/error-reports.js:
```diff
@@ -0,0 +1,105 @@
+        locationId,
+        services,
+        content,
+        postedBy, // not currently collected in frontend
```
I wouldn't expect us to necessarily ever collect these, and more generally, I don't think there's a need to add fields until we know they have some use. So, I'd just remove postedBy and contactInfo from all these places for now.

Rovack commented on src/controllers/error-reports.js:
```diff
@@ -0,0 +1,105 @@
+        throw new NotFoundError('Error report not found when attempting to delete it!');
+      }
+
+      // Authentication of administrative user
```
Here I do think the admin check makes sense.
However, I don't think the comment adds much that isn't obvious from the following line. (Which is good: readable code trumps comments any time.)

Rovack commented on src/models/error-report.js:
```diff
@@ -0,0 +1,36 @@
+  });
+
+  ErrorReport.associate = (models) => {
+    ErrorReport.belongsTo(models.Location);
```
The Location model's `associate` also needs to be updated.
Without that, I'm actually surprised if `location.createErrorReport` works. Have you tested it yet (through Postman or something)?
If not, might be easiest to add a test for creating error reports (similar to the existing add-comments.test.js).

cadecameron replied:
Oh yeah, I'd updated the location.js model to add the line:
```
Location.hasMany(models.ErrorReport);
```

Due to the PostGIS function name changing after version 2.2.0, I'd also changed one of the function names (from `'ST_Distance_Sphere'` to `'ST_DistanceSphere'`), and hence didn't stage the file.

I'll make sure this change gets in the next commit I push to this feature branch though.

Haven't even thought about the tests yet - thanks for pointing that out! I'll have a first pass at incorporating one.

Rovack commented on src/services/slack-notifier.js:
```diff
@@ -29,11 +29,46 @@ const notifyComment = (baseText, {
     ...commentParams
   }) => notifyComment(`New reply to comment "${originalComment.content}"`, commentParams),
+
+  notifyNewErrorReport: async errorReportParams => notifyErrorReport('New error report', errorReportParams),
```
Do you see a benefit to the split between `notifyNewErrorReport` and `notifyErrorReport`?
Looks like comments had a similar split, but that was because there was more than 1 type of event that could be reported. With error reports, it inherently seems like there can only be 1.

cadecameron replied:
> Do you see a benefit to the split between `notifyNewErrorReport` and `notifyErrorReport`?
> Looks like comments had a similar split, but that was because there was more than 1 type of event that could be reported. With error reports, it inherently seems like there can only be 1.

Good point. Thanks for picking up on this. 👍

> Unrelatedly, just remembered: right now the Slack channel is called something like "user-comments". Not sure if we should just rename it to something more general, or send these to a separate channel.
> What would it take if we wanted to configure a different channel here?

Yeah, I think a new channel in Slack would be beneficial.
I'll take a look at the docs for the Slack notification config.

cadecameron commented:
@Rovack, implemented the new 'general' column in the Postgres table, and updated necessary controllers/models/validation.

Any feedback, specifically on the error report controller, is welcomed. 👍

Rovack commented on src/controllers/error-reports.js:
```diff
@@ -0,0 +1,111 @@
+      // Parse certain values before sending to 'createInstance' in data-changes.js
+      // Perhaps find more elequent of checking for null values?
+      const services = req.body.services
+        ? JSON.parse(req.body.services)
```
How come this needs to be parsed? Looking at the validator, it looks like this is already an array.
If this was just meant to replace null values with an empty array, it could just be something like:
```
const services = req.body.services || [];
```
Though even that seems unnecessary; You could just make `services` required, and make sure to send at least an explicit empty array from the client.

Rovack commented on src/models/location.js:
```diff
@@ -41,6 +41,7 @@ module.exports = (sequelize, DataTypes) => {
     Location.hasMany(models.AccessibilityForDisabilities);
     Location.hasMany(models.EventRelatedInfo);
     Location.hasMany(models.Comment);
+    Location.hasMany(models.ErrorReport); // ADDED FOR ERROR REPORTS BY cade_cameron
```
No need for the comment. In the past, comments like these could sometimes be helpful for provenance, but nowadays we have source control for that (`git blame` shows this information for every line, plus the specific commit in which it was added, etc).

As a general sidenote, there seems to be a growing consensus in recent years that comments often do more harm than good. I've even seen teams outright ban comments from the codebase.
I think that's a bit extreme, but do agree that there's value in thinking twice before adding a comment, to consider whether that's the best tool to convey certain information, or whether it's symptomatic of code that could be made more self-explanatory, or generally could be better achieved some other way.

cadecameron replied:
Agree with the comments. Since I've had to make a local change to this location model (to accommodate the PostGIS function name change), it was more a marker for me that this change had to stay before making a commit. I've taken it out for the next commit though.

Rovack commented on src/services/slack-notifier.js:
```diff
@@ -29,11 +29,38 @@ const notifyComment = (baseText, {
+  }
+
+  if (services.length) {
+    text += `Services reported: ${services})\n`;
```
Having looked at the test message on Slack, it occurs to me that seeing the service IDs wouldn't mean much to anyone who doesn't have access to the DB, and anyone who does could just as easily pull up the full error report record.

What _would_ make it useful is if the controller fetched the actual service records from the DB and sent them to the notifier, which could then display the actual service names.

However, unless you think implementing that would be really quick and easy, it's probably not worth the effort at this point.
Could just post the content and org name on their own, and if any further info is needed, we can manually get it from the DB (and, of course, just display it in the admin UI once one is built).

Rovack replied:
(Of course, at that point showing the line about general location information is also probably not that useful, and this whole function can be like 3 lines.)

cadecameron replied:
Yeah the current service ids mean little. I figured that we'd either drop the line completely, or else provide the service name. I wasn't sure how much wrangling would be involved in getting the names though. Happy to drop the service ids out in the meantime (along with the general info).

cadecameron commented on src/controllers/error-reports.js:
```diff
@@ -0,0 +1,111 @@
+        throw new NotFoundError('Error report not found when attempting to delete it!');
+      }
+
+      if (!req.userIsAdmin) {
```
Good point - the auth should definitely come before the location lookup attempt. 😎👍

cadecameron commented on src/controllers/error-reports.js:
```diff
@@ -0,0 +1,111 @@
+      // Parse certain values before sending to 'createInstance' in data-changes.js
+      // Perhaps find more elequent of checking for null values?
+      const services = req.body.services
+        ? JSON.parse(req.body.services)
```
Yeah, there was some weird stuff happening with the bodyParser middleware (it was leaving the values as strings, rather than array or boolean values).

It might have been related to how I was testing in Postman though, so I'll investigate this issue in conjunction with the frontend.

cadecameron replied:
The `req.body` values have been tested with the frontend components, and have been updated.

Rovack commented:
@cadecameron Everything looks great now.
Just let me know when it's ready (e.g. "Remove following comments after testing" comment is removed), and we can merge this.

Rovack commented on src/controllers/validation/error-report.js:
```diff
@@ -10,9 +10,9 @@ export default {
-      content: Joi.string(),
+      generalLocationError: Joi.boolean().required(),
+      services: Joi.array().items(Joi.string().guid()).required(),
+      content: Joi.string().empty(''),
```
If content was skipped, wouldn't it be better to send a null value?

cadecameron replied:
We could exclude `content` from the frontend POST request, in which case I'd make `content` in the Joi validation file optional, and use a default value of `null` perhaps?

Rovack replied:
Yeah. It's already not marked as required in the DB nor here, so should just be a matter of removing `.empty('')` from this line.
